src/atoms/affine/dot.jl

```diff
@@ -1,26 +1,19 @@
 import LinearAlgebra.dot
-export vecdot, dot
+export dot
 
+ismatrix(x::AbstractExpr) = (s = size(x); length(s) == 2 && !any(==(1), s))
+ismatrix(x::AbstractMatrix) = true
+ismatrix(x) = false
 
-vecdot(x::AbstractExpr, y::AbstractExpr) = sum(broadcast(*, x, y))
-vecdot(x::Value, y::AbstractExpr) = sum(broadcast(*, Constant(x), y))
-vecdot(x::AbstractExpr, y::Value) = sum(broadcast(*, x, Constant(y)))
+asvec(x) = ismatrix(x) ? convert(AbstractExpr, vec(x)) : convert(AbstractExpr, x)
 
-dot(x::AbstractExpr, y::AbstractExpr) = (ismatrix(x) || ismatrix(y)) ? error("dot not implemented for matrices. perhaps you're looking for vecdot?") : vecdot(x, y)
-dot(x::Value, y::AbstractExpr) = (ismatrix(x) || ismatrix(y)) ? error("dot not implemented for matrices. perhaps you're looking for vecdot?") : vecdot(x, y)
-dot(x::AbstractExpr, y::Value) = (ismatrix(x) || ismatrix(y)) ? error("dot not implemented for matrices. perhaps you're looking for vecdot?") : vecdot(x, y)
+dot(x::AbstractExpr, y::AbstractExpr) = sum(broadcast(*, asvec(x), asvec(y)))
```

Comment:
Can we do this without `asvec`? That works on normal matrices. Or are you trying to ensure length is the same to avoid extending singleton dimensions?

Reply:
I don't really know what I was doing. I made this PR right before I went to bed last night. Most of this logic is completely unnecessary.

Reply:
Oh wait, yes, the lengths need to be the same and we need to avoid broadcast-specific behaviors